src/ring_theory/norm.lean

```diff
@@ -0,0 +1,102 @@
+/-
+Copyright (c) 2021 Anne Baanen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anne Baanen
+-/
+
+import linear_algebra.determinant
+import ring_theory.power_basis
+
+/-!
+# Norm for (finite) ring extensions
+
+Suppose we have an `R`-algebra `S` with a finite basis. For each `s : S`,
+the determinant of the linear map given by multiplying by `s` gives information
+about the roots of the minimal polynomial of `s` over `R`.
+
+## Implementation notes
+
+Typically, the norm is defined specifically for finite field extensions.
+The current definition is as general as possible and the assumption that we have
+fields or that the extension is finite is added to the lemmas as needed.
+
+We only define the norm for left multiplication (`algebra.left_mul_matrix`,
+i.e. `algebra.lmul_left`).
+For now, the definitions assume `S` is commutative, so the choice doesn't
+matter anyway.
+
+See also `algebra.trace`, which is defined similarly as the trace of
+`algebra.left_mul_matrix`.
+
+## References
+
+ * https://en.wikipedia.org/wiki/Field_norm
+
+-/
+
+universes u v w
+
+variables {R S T : Type*} [integral_domain R] [integral_domain S] [integral_domain T]
+variables [algebra R S] [algebra R T]
+variables {K L : Type*} [field K] [field L] [algebra K L]
+variables {ι : Type w} [fintype ι]
+
+open finite_dimensional
+open linear_map
+open matrix
+
+open_locale big_operators
+open_locale matrix
+
+namespace algebra
+
+variables (b : basis ι R S)
+
+variables (R S)
+
+/-- The norm of an element `s` of an `R`-algebra is the determinant of `(*) s`. -/
+noncomputable def norm : S →* R :=
```

Comment:
Should `S` be explicit? My intuition is that in 90% of the cases it can be inferred from `x`.

Reply:
The original reason for having it explicit is to match `lmul R S`, `algebra_map R S`, ... It's not needed in any of the code I wrote up til now, so we can safely implicitify it. (And if it's needed, we can always write `norm R : R →+* S`.)

Reply:
Back in the old days Lean couldn't infer `S` from `x`, but then Gabriel stepped in and made Lean smarter. I think `algebra_map` and friends haven't been updated yet.